### PR TITLE
Add scala version.

### PR DESCRIPTION
**Pull request by giiita**
## Add scala version
https://github.com/frol/completely-unscientific-benchmarks/issues/25

## Test
スクリーンショット 2020-04-16 22 28 19

### CODE REVIEW

**Comment by giiita** [comment]
@frol Please review it.

**Comment by frol** [comment]
@giiita I am sorry, I was oversubscribed with the other items on my table.

**Comment by frol** [comment]
@sivakon @FranklinChen Can you help to review the suggested solution? I want to remind you that you don't want to run and see the performance characteristics. To qualify for the naive solution you should write the code the way you would write it the very first time, so the only question is: does it look like something you would write in idiomatic Scala?

File: scala/Main.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    new SplitResult(lowerOther.first, equalGreater.first, equalGreater.second)
+  }
+}
+
+case class Node(var x: Int, var left: Node = null, var right: Node = null) {
+  val y: Int = Node.random.nextInt
+}
+
+class Tree(var mRoot: Node = null) {
+  def hasValue(x: Int, res: Int): (Int, Tree) = {
```
Comment (frol) [inline scala/Main.scala]: This method should just return boolean, unless it is uncommon to use booleans in Scala (is it the case?)

File: scala/Main.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  def insert(x: Int): Tree = {
+    val splited = Node.split(mRoot, x)
+    if (splited.equal == null) {
+      splited.equal = new Node(x)
+    }
+    mRoot = Node.merge(splited.lower, splited.equal, splited.greater)
+    this
+  }
+
+  def erase(x: Int): Tree = {
```
Comment (frol) [inline scala/Main.scala]: Can it return "nothing"? I wonder if you followed some existing implementation where we return a node instead of nothing in `erase` method :thinking:

File: scala/Makefile [hunk 3]
```diff
@@ -0,0 +1,19 @@
+.PHONY: all
```
Comment (frol) [inline scala/Makefile:1]: Thank you for including the Makefile!

**Review by frol — APPROVED** [review]
Overall, looks good to me. Yet, I have never written a single line of Scala code, so I cannot tell if it is idiomatic enough (I find it straightforward, but I don't know where to use `class` over `object`, for example)

**Comment by giiita** [comment]
@frol 
Scala is a functional language. Therefore, if you follow Scala's theory instead of Java like, can use

- Do not use null
- Don't use var
- Maintain reference transparency

These properties need to be maintained. However, when following this testing algorithm, the overhead of being immutable is quite large, and even in Scala, it is better to implement complex algorithms as Java-like as possible, as the performance will be much better.

In other words, for this benchmark, the best thing to do would be to write like Java. However, in many cases that's not what Scala wants, so we rewrote it to be immutable and reference transparent.
If you want high performance, you can revert to the latest commit.

Sorry for the late response, but I want to close this PR asap. Please merge, revert and merge, or close.

**Comment by frol** [comment]
Thank you for the contribution! I will review it later this week.

**Comment by giiita** [comment]
I'll close this once.
